pubsub-mapped-api/src/main/java/com/google/pubsub/clients/consumer/KafkaConsumer.java

```diff
@@ -0,0 +1,592 @@
+  }
+
+  private ListenableFuture<Subscription> deputeSinglePubsubSubscription(String subscriptionString,
+      String topicName) {
+    SubscriberFutureStub subscriberFutureStub = stubCreator.getSubscriberFutureStub();
+    return subscriberFutureStub
+        .createSubscription(Subscription.newBuilder()
+            .setName(subscriptionString)
+            .setTopic(TOPIC_PREFIX + topicName)
+            .build());
+  }
+
+  private Map<String, Subscription> getNewSubscriptionMap(
+      List<ResponseData<Subscription>> responseDatas) {
+    Map<String, Subscription> subscriptionMap = new HashMap<>();
+
+    for(ResponseData<Subscription> responseData: responseDatas) {
+      try {
+        Subscription s = responseData.getRequestListenableFuture().get();
+        subscriptionMap.put(responseData.getTopicName(), s);
+      } catch (InterruptedException e) {
+        //if an error is thrown, attempt to delete subscriptions created in this loop
+        unsubscribe(subscriptionMap.values());
+        throw new InterruptException(e);
+      } catch (ExecutionException e) {
+        unsubscribe(subscriptionMap.values());
+        throw new KafkaException(e);
+      }
+    }
+    return subscriptionMap;
+  }
+
+  private void checkSubscribePreconditions(Collection<String> topics) {
+    if(topics == null) {
+      throw new IllegalArgumentException("Topic collection to subscribe to cannot be null");
+    }
+
+    for(String topic: topics) {
+      if(topic == null || topic.trim().isEmpty())
+        throw new IllegalArgumentException(
+            "Topic collection to subscribe to cannot contain null or empty topic");
+    }
+  }
+
+  @Override
+  public void subscribe(Collection<String> topics) {
+    subscribe(topics, null);
+  }
+
+  /*
+  Subscribe is pulling all topics from PubSub project and subscribing to ones matching provided
+  pattern.
+   */
+  @Override
+  public void subscribe(Pattern pattern, ConsumerRebalanceListener listener) {
+    checkPatternSubscribePreconditions(pattern);
+
+    List<Topic> existingTopics = getPubsubExistionTopics();
+
+    List<String> matchingTopics = new ArrayList<>();
+
+    for(Topic topic: existingTopics) {
+      String topicName = topic.getName().substring(TOPIC_PREFIX.length(), topic.getName().length());
+      Matcher m = pattern.matcher(topicName);
+      if(m.matches())
+        matchingTopics.add(topicName);
+    }
+    subscribe(matchingTopics);
+
+    log.debug("Subscribed to pattern: {}", pattern);
+  }
+
+  private List<Topic> getPubsubExistionTopics() {
+    PublisherBlockingStub publisherBlockingStub = stubCreator.getPublisherBlockingStub();
+
+    ListTopicsResponse listTopicsResponse = publisherBlockingStub
+        .listTopics(ListTopicsRequest.newBuilder()
+            .setProject(GOOGLE_CLOUD_PROJECT)
+            .build());
+
+    return listTopicsResponse.getTopicsList();
+  }
+
+  private void checkPatternSubscribePreconditions(Pattern pattern) {
+    if(pattern == null) {
```

Comment:
yes.  

That said, expecting IAE instead of NPE when passing a null is very surprising.